### ISSUE TITLE
Does this package have a roadmap for better devops features ?

### ISSUE THREAD

**Issue by fcharras** [NVIDIA/apex#381]
This package so far does not play well with healthy deployment pipelines:
- provides no versionning
- does not run on CPU, which is practical for gpu-less CI (what's the reason for this limitation ?)
- if not for official releases, automated run of the tests on github could reassure about the stability but there's not that either.

As a python user I've also failed so far to make AMP work smoothly with non trivial scenarii (switching optimizer seems to create memory leaks, pickling will not work, seemingly because of monkey patching with locally defined functions,...). While this might be my mistake, debugging looks excessively complicated because of the globals like `_DECORATOR_HANDLE` and `_amp_state.handle`. Such global variables (whose state are changed by functions as side effects) looks like bad development practices. If the use of said variables are inevitable the contribution guidelines should state it in and methods should be provided to reset the state of the package as if it was freshly imported to avoid leaks. 

Or if this package is purely for POCs and linear research scripts, it should be flagged as such and state clearly that other usecases are not supported ?

Nonetheless this package seems to be marketed like a mature feature by Nvidia ( https://developer.nvidia.com/automatic-mixed-precision ), has by now 1800 stars on github, and the repo itself is hosted by the NVIDIA group, so people could expect, if not a mature package by usual standards, an acknowledgement that this work is at most an alpha release, and a draft of roadmap on the afore-mentionned features before a first stable release.

**Comment by mcarilli** [comment]
Amp began as a research tool and continues to serve mainly in that capacity.  We've had one (and in the past two months two) people working on it part time and it has been a learning process.  One of the biggest pain points is that people's environments often aren't set up to build cpp extensions, which isn't as easy as `pip install torch`.  We have a short-term plan to provide pip packages, but then people must match the versions for those as well.

To avoid install issues, and other issues you note, the roadmap is to move Amp functionality into Pytorch core over the next few months, into both the C++ backend (for eager mode execution) and also as a jit pass for tracing/scripting.  The existing Apex repo will continue to be supported for people who are already using it.

Fixes and changes are included often enough, and the install process is quick enough (when extension building works) that it hasn't made sense to provide numbered releases.  We run tests locally on Pytorch versions back to 1.0 and installing from the latest Apex master is always best.

CPU support is pointless, the op coverage for half precision on CPU in Pytorch is significantly less than the GPU coverage, so there's nothing we can do.

Monkey-patching with locally defined functions is the only way to make Amp work without getting code in upstream; the planned upstream implementation will not require any Python-side monkey patching.

> switching optimizer seems to create memory leaks
I don't know what you mean by this but that at least sounds fixable.  What exactly are you trying to do?

**Comment by fcharras** [comment]
Thank you for the insight, I would love to see this added upstream.

> installing from the latest Apex master is always best
said no devops ever :grin: it would make more sense to call the project a WIP, a POC or an alpha for clarity.

> Monkey-patching with locally defined functions is the only way to make Amp work without getting code in upstream
I'm not very knowledgeable of monkey-patching but consider the following snippet: (note: joblib is basically a wrapper on pickle, used it there out of habit)
```
import joblib

# Let's create a class and an instance of this class that are picklables
class A:
    def func(self):
        return 0
a = A()
joblib.dump(a, './test.pkl')

# mock method with an outer function and check that it can be pickled
def global_func(self):
    return 1
def mock_global(a):
    a.func = global_func
    return a
a = mock_global(a)
joblib.dump(a, './test.pkl')

# mock with an inner function breaks pickling
def mock_local(a):
    def local_func(self):
        return 2
    a.func = local_func
    return a
a = A()
a = mock_local(a)
joblib.dump(a, './test.pkl')  # error
```
If you replace all the inner functions with outer functions it might repair pickling of monkey-patched instances ? (I haven't investigated more thorougly on this issue but at least this minimal example makes sense right ?)

> switching optimizer seems to create memory leaks
I'll provide more information if I manage to fix it, it does sound fixable, my point was rather that globals make it harder.

**Comment by mcarilli** [comment]
I'm still not sure what exactly you mean by "switching optimizers".  I might be able to provide guidance if I have a clearer idea.  Do you mean you start training with one optimizer and then later change to another after some epochs?

**Comment by fcharras** [comment]
Yes, for some reason I tried to reset an optimizer at each epoch (by getting a new fresh instance). I see the vram slowly increasing after each reset until a memory error. To be honest I don't have a minimal example that clearly show that amp is the cause, I suspected that some references to tensors might keep existing in the globals but that might be something else.
